Frameworks/Foundation/NSOperation.mm

```diff
@@ -111,32 +115,36 @@ - (id)init {
  @Status Interoperable
 */
 - (BOOL)isReady {
-    std::lock_guard<std::recursive_mutex> lock(_readyLock);
+    std::lock_guard<std::recursive_mutex> lock(_dependenciesLock);
     return _ready;
 }
 
 /**
  @Status Interoperable
 */
 - (void)addDependency:(id)operation {
-    std::lock_guard<std::recursive_mutex> lock(_dependenciesLock);
+    std::lock_guard<std::recursive_mutex> dependenciesLock(_dependenciesLock);
     [self willChangeValueForKey:@"dependencies"];
     [_dependencies addObject:operation];
-    [operation addObserver:self forKeyPath:@"isFinished" options:0 context:(void*)NSOperationContext];
+    [operation addObserver:self forKeyPath:@"isFinished" options:NSKeyValueObservingOptionInitial context:(void*)NSOperationContext];
     [self didChangeValueForKey:@"dependencies"];
+    _outstandingDependenciesCount++; // will be decremented if necessary in observeValueForKey, due to NSKeyValueObservingOptionInitial
```

Comment:
>_outstandingDependenciesCount++; // will be decremented if necessary in observeValueForKey, due to NSKeyValueObservingOptionInitial [](start = 4, length = 131)

seems wrong. What if dependency is already finished? Shouldn't you only increment for unfinished ones? Does KVO send an immediate observation for the current state or only when it changes? Additionally this logic seems wrong if you add the same dependency twice. You will send a change notification for _dependencies when no change occured and you'll be double counting the same object which is unclear how the _outstandingDepCount will work out....